solidity/ecdsa/test/WalletRegistry.Rewards.test.ts

```diff
@@ -0,0 +1,174 @@
+
+import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import type { FakeContract } from "@defi-wonderland/smock"
+import type { Operator, OperatorID } from "./utils/operators"
+import type {
+  SortitionPool,
+  WalletRegistry,
+  WalletRegistryGovernance,
+  TokenStaking,
+  IWalletOwner,
+  T,
+} from "../typechain"
+
+const { to1e18 } = helpers.number
+
+const { createSnapshot, restoreSnapshot } = helpers.snapshot
+
+describe("WalletRegistry - Rewards", () => {
+  let tToken: T
+  let walletRegistry: WalletRegistry
+  let staking: TokenStaking
+  let walletRegistryGovernance: WalletRegistryGovernance
+  let sortitionPool: SortitionPool
+  let walletOwner: FakeContract<IWalletOwner>
+
+  let deployer: SignerWithAddress
+  let governance: SignerWithAddress
+  let thirdParty: SignerWithAddress
+
+  let members: Operator[]
+  let membersIDs: OperatorID[]
+  let walletID: string
+
+  const walletPublicKey: string = ecdsaData.group1.publicKey
+
+  const rewardAmount = to1e18(100000)
+
+  before("load test fixture", async () => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;({
+      tToken,
+      walletRegistry,
+      walletRegistryGovernance,
+      staking,
+      sortitionPool,
+      walletOwner,
+      deployer,
+      governance,
+      thirdParty,
+    } = await walletRegistryFixture())
+    ;({ members, walletID } = await createNewWallet(
+      walletRegistry,
+      walletOwner.wallet,
+      walletPublicKey
+    ))
+
+    membersIDs = members.map((member) => member.id)
+  })
+
+  describe("withdrawRewards", () => {
+    context("when called for an unknown operator", () => {
+      it("should revert", async () => {
+        await expect(
+          walletRegistry.withdrawRewards(thirdParty.address)
+        ).to.be.revertedWith("Unknown operator")
+      })
+    })
+
+    context("when called for a known operator", () => {
+      before(async () => {
+        await createSnapshot()
+
+        // Allocate sortition pool rewards
+        await tToken.connect(deployer).mint(deployer.address, rewardAmount)
+        await tToken
+          .connect(deployer)
+          .approveAndCall(sortitionPool.address, rewardAmount, [])
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should withdraw rewards", async () => {
+        const operator = members[0].signer.address
+        const stakingProvider = await walletRegistry.operatorToStakingProvider(
+          operator
+        )
+        const { beneficiary } = await staking.rolesOf(stakingProvider)
+
+        expect(await tToken.balanceOf(beneficiary)).to.equal(0)
+        await walletRegistry.withdrawRewards(stakingProvider)
+        expect(await tToken.balanceOf(beneficiary)).to.be.gt(0)
+      })
+    })
+  })
+
+  describe("withdrawIneligibleRewards", () => {
+    const inactiveMembersIndices = [1, 5, 10]
+    const heartbeatFailed = false
```

Comment:
Not biggy, but now we sort of have 2 phrases to describe the same thing. Before we only used "heartbeat failures" and then we started using "operator inactivity". How about we rename it to "isOperatorActive"?

Reply:
It is not the same thing. Both `WalletRegistry` and `RandomBeacon` has a function to `notifyOperatorInactivity`. In both cases, this function is checking signatures of group members supporting the claim, and marking inactive operators as not eligible for rewards.

Only `WalletRegistry` accepts an additional parameter to the inactivity claim: a boolean flag whether the heartbeat failed:
https://github.com/keep-network/keep-core/blob/8dc2df00236d2a09f038187c83fd8441aa731d63/solidity/ecdsa/contracts/libraries/EcdsaInactivity.sol#L40-L50

Operator inactivity is a minor event leading to marking operators as ineligible for rewards.

Failed heartbeat (due to too many inactive operators) is a major event and leads to notification of tBTC `Bridge` that triggers the procedure of moving funds to another ECDSA wallet.

Reply:
Ah, I see.. so we use "operator inactivity" when talking about a single operator, but "failed heartbeat" when many.